src/include/daos_pipeline.h

```diff
@@ -0,0 +1,261 @@
+/**
+ * (C) Copyright 2021 Intel Corporation.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause-Patent
+ */
+
+#ifndef __DAOS_PIPE_H__
+#define __DAOS_PIPE_H__
+
+#if defined(__cplusplus)
+extern "C" {
+#endif
+
+
+/** logic functions */
+#define DAOS_FILTER_FUNC_EQ		0
+#define DAOS_FILTER_FUNC_NE		1
+#define DAOS_FILTER_FUNC_LT		2
+#define DAOS_FILTER_FUNC_LE		3
+#define DAOS_FILTER_FUNC_GE		4
+#define DAOS_FILTER_FUNC_GT		5
+#define DAOS_FILTER_FUNC_LIKE		6
+#define DAOS_FILTER_FUNC_ISNULL		7
+#define DAOS_FILTER_FUNC_ISNOTNULL	8
+#define DAOS_FILTER_FUNC_AND		9
+#define DAOS_FILTER_FUNC_OR		10
+
+/** aggregation functions */
+#define DAOS_FILTER_FUNC_SUM		100
+#define DAOS_FILTER_FUNC_MIN		101
+#define DAOS_FILTER_FUNC_MAX		102
+#define DAOS_FILTER_FUNC_AVG		103
+
+/** keys, constants */
+#define DAOS_FILTER_DKEY		200
+#define DAOS_FILTER_AKEY		201
+#define DAOS_FILTER_CONST		202
+
+/** types used in a filter object */
+#define DAOS_FILTER_TYPE_BINARY		0
+#define DAOS_FILTER_TYPE_STRING		1
+#define DAOS_FILTER_TYPE_INTEGER	2
+#define DAOS_FILTER_TYPE_REAL		3
+
+/** types of pipeline nodes */
+#define DAOS_PIPELINE_CONDITION		0
+#define DAOS_PIPELINE_AGGREGATION	1
+
+
+/**
+ * A filter object, used to build operations for a pipeline node.
+ *
+ */
+typedef struct {
+	/**
+	 * filter type can be any of the following:
+	 *   -- functions:
+	 *      - logical functions:
+	 *          DAOS_FILTER_FUNC_EQ:		==
+	 *          DAOS_FILTER_FUNC_NE:		!=
+	 *          DAOS_FILTER_FUNC_LT:		<
+	 *          DAOS_FILTER_FUNC_LE:		<=
+	 *          DAOS_FILTER_FUNC_GE:		>=
+	 *          DAOS_FILTER_FUNC_GT:		>
+	 *          DAOS_FILTER_FUNC_LIKE:		== (reg exp.)
+	 *          DAOS_FILTER_FUNC_ISNULL:		==NULL
+	 *          DAOS_FILTER_FUNC_ISNOTNULL:		!=NULL
+	 *          DAOS_FILTER_FUNC_AND:		&&
+	 *          DAOS_FILTER_FUNC_OR:		||
+	 *      - aggeration functions:
+	 *          DAOS_FILTER_FUNC_SUM:		SUM()
+	 *          DAOS_FILTER_FUNC_MIN:		MIN()
+	 *          DAOS_FILTER_FUNC_MAX:		MAX()
+	 *          DAOS_FILTER_FUNC_AVG:		AVG()
+	 *   -- key:
+	 *          DAOS_FILTER_DKEY:	Filter object represents dkey
+	 *          DAOS_FILTER_AKEY	Filter object represents akey
+	 *   -- constant:
+	 *          DAOS_FILTER_CONST:	Filter object is a constant
+	 */
+	int		filter_type;
+	/**
+	 * Type of data. Only relevant for keys and constant filter type objects:
+	 *          DAOS_FILTER_TYPE_BINARY
+	 *          DAOS_FILTER_TYPE_STRING
+	 *          DAOS_FILTER_TYPE_INTEGER
+	 *          DAOS_FILTER_TYPE_REAL
+	 */
+	int		data_type;
+	/**
+	 * Number of parameters for this filter object. For example, for '=='
+	 * we have 2 parameters.
+	 */
+	uint32_t	num_params;
+	/**
+	 * If filtering by akey, this tells us which one.
+	 */
+	d_iov_t		akey;
```

Comment:
So the actual filter applies on the dkey only?

Reply:
No. We have to explicitly pass the akey, because there can be multiple akeys (if, for example, you filter by "age<4", so have to pass "age"). But if a filter just operates on the dkey, there is only one (if we filter by dkey, we just say dkey<4, no need to pass which one). To filter by dkey we define the type of the filter as DAOS_FILTER_DKEY.